pkg/systemd/services/services.jsx

```diff
@@ -537,202 +522,54 @@
-            needsUpdate = true;
-        }
-        const system_boot_time = clock_realtime_now * 1000 - clock_monotonic_now;
-        if (timer_unit.LastTriggerUSec === -1 || timer_unit.LastTriggerUSec === 0) {
-            if (unit.LastTriggerTime !== _("unknown")) {
-                unit.LastTriggerTime = _("unknown");
-                needsUpdate = true;
-            }
-        }
-        let next_run_time = 0;
-
+    addTimerProperties(timer_props, unit) {
+        const last_trigger_usec = timer_props.LastTriggerUSec.v;
         // systemd puts -1 into an unsigned int type for the various *USec* properties
         // JS rounds these to a float which is > MAX_UINT64, but the comparison works
-        if (timer_unit.NextElapseUSecRealtime > 0 && timer_unit.NextElapseUSecRealtime < MAX_UINT64) {
-            next_run_time = timer_unit.NextElapseUSecRealtime;
-        } else if (timer_unit.NextElapseUSecMonotonic > 0 && timer_unit.NextElapseUSecMonotonic < MAX_UINT64) {
-            next_run_time = timer_unit.NextElapseUSecMonotonic + system_boot_time;
-        }
-
-        const nextRunTime = timeformat.dateTime(next_run_time / 1000);
-        if (nextRunTime !== unit.NextRunTime) {
-            unit.NextRunTime = nextRunTime;
-            needsUpdate = true;
-        }
-
-        if (next_run_time === 0 && unit.NextRunTime !== _("unknown")) {
-            unit.NextRunTime = _("unknown");
-            needsUpdate = true;
-        }
+        if (last_trigger_usec > 0 && last_trigger_usec < MAX_UINT64)
+            unit.LastTriggerTime = timeformat.dateTime(last_trigger_usec / 1000);
+        else
+            unit.LastTriggerTime = _("unknown");
 
-        return needsUpdate;
+        const system_boot_time = clock_realtime_now * 1000 - clock_monotonic_now;
+        const next_realtime = timer_props.NextElapseUSecRealtime?.v;
+        const next_monotonic = timer_props.NextElapseUSecMonotonic?.v;
+        let next_run_time = null;
+        if (next_realtime > 0 && next_realtime < MAX_UINT64)
+            next_run_time = next_realtime;
+        else if (next_monotonic > 0 && next_monotonic < MAX_UINT64)
+            next_run_time = next_monotonic + system_boot_time;
+
+        unit.NextRunTime = next_run_time ? timeformat.dateTime(next_run_time / 1000) : _("unknown");
     }
 
     /* Add some computed properties into a unit object - does not call setState */
     updateComputedProperties(unit) {
-        unit.HasFailed = (unit.ActiveState == "failed" || (unit.LoadState !== "loaded" && unit.LoadState != "masked"));
+        unit.HasFailed = unit.ActiveState == "failed" || (
+            unit.LoadState && unit.LoadState !== "loaded" && unit.LoadState !== "masked");
 
         unit.CombinedState = this.activeState[unit.ActiveState] || unit.ActiveState;
-        if (unit.LoadState !== "loaded" && unit.LoadState != "masked")
+        if (unit.LoadState && unit.LoadState !== "loaded" && unit.LoadState !== "masked")
```

Comment:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18422-20230307-120556-7e7cb6b6-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)